 <h1 style="text-align: center; color: RED">MÜŞTERİ YAŞAM BOYU DEĞERİ</h1> 

Bir müşterinin bir şirketle kurduğu ilişki-iletişim süresince bu şirkete kazandıracağı parasal değerdir. Müşterilerin gelecekte de bize sağlayacağı faydayı belirleyebilirsek buna göre hem onlarla olan ilişklerimizi düzenleyebilir hem de şirketimizde orta uzun vadeli daha müşteri odaklı bir yaklaşım sergileyebiliriz.Mevcut müşteri kitlesi ile yeni müşteri bulma maliyetini kıyaslama imkanı bulabiliriz. 

<span style="color: blue; font-weight: bold;">Nasıl Hesaplanır?</span> 

satın alma başına ortalama kazanç * satın alma sayısı

**CLTV** = (Customer Value / Churn Rate) * Profit Margin

* Churn Rate: Müşteri terk oranıdır ve bu bir sabittir.
* Profit Margin: Şirketin müşteriler ile yaptığı alışverişlerde varsayacağı kar miktarıdır.

Customer Value = Average Order Value * Purchase Frequency

* Purchase Frequency: Satın alma frekansıdır.

Average Order Value= Total Price / Total Transaction

İşlem başına bırakılan kazanç demektir.

Purchase Frequency = Total Transaction / Total Number of Customer

Toplam işlem sayının toplam müşteri sayısına oranıdır.

Churn Rate = 1- Repaet Rate
* Repeat Rate: Tekrar oranı, elde tutma oranıdır.

Repaet Rate = Birden fazla alışveriş yapan müşteri sayısı / tüm müşteriler

Profit Margin = Total Price * 0.10

Sonuç olarak her bir müşteri için hesaplanacak olan CLTV değerlerine göre bir sıralama yapıldığında ve CLTV değerlerine göre belirli noktalardan bölme işlemi yapılarak gruplar oluşturulduğunda müşterilerimiz segmentlere ayrılmış olucak.

 <h1 style="text-align: center; color: RED">UYGULAMA</h1>

1. Veri Hazırlama
2. Average Order Value (average_order_value = total_price / total_transaction)
3. Purchase Frequency (total_transaction / total_number_of_customers)
4. Repeat Rate & Churn Rate (birden fazla alışveriş yapan müşteri sayısı / tüm müşteriler)
5. Profit Margin (profit_margin =  total_price * 0.10)
6. Customer Value (customer_value = average_order_value * purchase_frequency)
7. Customer Lifetime Value (CLTV = (customer_value / churn_rate) x profit_margin)
8. Segmentlerin Oluşturulması
9. BONUS: Tüm İşlemlerin Fonksiyonlaştırılması

<span style="color: blue; font-weight: bold;">1. VERİ HAZIRLAMA</span> 

**Veri Seti Hikayesi**


Online Retail II isimli veri seti İngiltere merkezli online bir satış mağazasının 01/12/2009 - 09/12/2011 tarihleri arasındaki satışlarını içeriyor.

**Değişkenler**

InvoiceNo: Fatura numarası. Her işleme yani faturaya ait eşsiz numara. C ile başlıyorsa iptal edilen işlem.

StockCode: Ürün kodu. Her bir ürün için eşsiz numara.

Description: Ürün ismi

Quantity: Ürün adedi. Faturalardaki ürünlerden kaçar tane satıldığını ifade etmektedir.

InvoiceDate: Fatura tarihi ve zamanı.

UnitPrice: Ürün fiyatı (Sterlin cinsinden)

CustomerID: Eşsiz müşteri numarası

Country: Ülke ismi. Müşterinin yaşadığı ülke.

Amacımız veri setindeki her bir müşteri için Customer Lifetime Value degerini hesaplamak
ve Lifetime Value üzerinden  segmentasyon yapmaktır.

In [29]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
pd.set_option('display.max_columns',None)
pd.set_option('display.width',500)
pd. set_option('display.float_format', lambda x:'%.5f' % x)


In [31]:
df_ = pd.read_excel("C:\\Users\\ASLI\\Desktop\\Data Scientist Bootcamp\\3.CRM Analytics\\resim\\online_retail_II.xlsx", sheet_name = 'Year 2009-2010')

In [33]:
df = df_.copy()

In [35]:
df.head()

Invoice StockCode                          Description  Quantity         InvoiceDate   Price  Customer ID         Country
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12 2009-12-01 07:45:00 6.95000  13085.00000  United Kingdom
1  489434    79323P                   PINK CHERRY LIGHTS        12 2009-12-01 07:45:00 6.75000  13085.00000  United Kingdom
2  489434    79323W                  WHITE CHERRY LIGHTS        12 2009-12-01 07:45:00 6.75000  13085.00000  United Kingdom
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48 2009-12-01 07:45:00 2.10000  13085.00000  United Kingdom
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24 2009-12-01 07:45:00 1.25000  13085.00000  United Kingdom

In [39]:
df.shape

(525461, 8)

In [37]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [43]:
#İade olan gözlemleri veri setinden çıkardık
df  = df [~df['Invoice'].str.contains('C', na=False)]

In [45]:
df.shape


(515255, 8)

In [47]:
df.describe().T

count                           mean                  min                  25%                  50%                  75%                  max        std
Quantity    515255.00000                       10.95663          -9600.00000              1.00000              3.00000             10.00000          19152.00000  104.35401
InvoiceDate       515255  2010-06-28 17:41:06.558422528  2009-12-01 07:45:00  2010-03-21 13:27:00  2010-07-06 13:13:00  2010-10-15 14:27:00  2010-12-09 20:01:00        NaN
Price       515255.00000                        3.95637         -53594.36000              1.25000              2.10000              4.21000          25111.09000  127.68856
Customer ID 407695.00000                    15368.50411          12346.00000          13997.00000          15321.00000          16812.00000          18287.00000 1679.79570

In [49]:
df= df[df['Quantity'] > 0]

In [51]:
df.describe().T

count                           mean                  min                  25%                  50%                  75%                  max        std
Quantity    513134.00000                       11.71543              1.00000              1.00000              3.00000             10.00000          19152.00000   92.97472
InvoiceDate       513134  2010-06-28 21:20:02.384289536  2009-12-01 07:45:00  2010-03-21 15:03:00  2010-07-06 16:26:00  2010-10-15 14:51:00  2010-12-09 20:01:00        NaN
Price       513134.00000                        3.97273         -53594.36000              1.25000              2.10000              4.21000          25111.09000  127.95193
Customer ID 407695.00000                    15368.50411          12346.00000          13997.00000          15321.00000          16812.00000          18287.00000 1679.79570

In [53]:
df.dropna(inplace=True)

In [55]:
df.describe().T


count                           mean                  min                  25%                  50%                  75%                  max        std
Quantity    407695.00000                       13.58669              1.00000              2.00000              5.00000             12.00000          19152.00000   96.84223
InvoiceDate       407695  2010-07-01 10:10:10.782177792  2009-12-01 07:45:00  2010-03-26 14:01:00  2010-07-09 15:46:00  2010-10-14 17:09:00  2010-12-09 20:01:00        NaN
Price       407695.00000                        3.29419              0.00000              1.25000              1.95000              3.75000          10953.50000   34.75666
Customer ID 407695.00000                    15368.50411          12346.00000          13997.00000          15321.00000          16812.00000          18287.00000 1679.79570

In [57]:
df['TotalPrice'] = df['Price'] * df['Quantity'] #toplam kazançlar veri setinde yoktu biz ekledik

In [59]:
df.head()

Invoice StockCode                          Description  Quantity         InvoiceDate   Price  Customer ID         Country  TotalPrice
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12 2009-12-01 07:45:00 6.95000  13085.00000  United Kingdom    83.40000
1  489434    79323P                   PINK CHERRY LIGHTS        12 2009-12-01 07:45:00 6.75000  13085.00000  United Kingdom    81.00000
2  489434    79323W                  WHITE CHERRY LIGHTS        12 2009-12-01 07:45:00 6.75000  13085.00000  United Kingdom    81.00000
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48 2009-12-01 07:45:00 2.10000  13085.00000  United Kingdom   100.80000
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24 2009-12-01 07:45:00 1.25000  13085.00000  United Kingdom    30.00000

* Veri setini clv hesaplanması için bize lazım olan metriklere göre çevirmemiz gerekmektedir.
Bu metriklere ulaşmak için müşteri özelinde total_price,
total_transaction(her birinin yaptığı toplam işlem) yani her fatura bizim için bir işlem
total_number_of_customer yani kişinin yaptığı toplam işlem
burda amacımız müşteriyi tekilleştirerek ınvoice yani total_transaction(toplam işlem)
ve toplam totalPrice ları görebilmek
böyle yaparak veri setini cltv hesaplması için uygun hale getirdik.

In [65]:
cltv_c = df.groupby('Customer ID').agg({'Invoice': lambda x: x.nunique(),#her müşterinin eşsiz faturası yani total transaction
                                        'Quantity': lambda x: x.sum(),
                                        'TotalPrice': lambda x: x.sum()})

In [67]:
cltv_c

Invoice  Quantity  TotalPrice
Customer ID                               
12346.00000       11        70   372.86000
12347.00000        2       828  1323.32000
12348.00000        1       373   222.16000
12349.00000        3       993  2671.14000
12351.00000        1       261   300.93000
...              ...       ...         ...
18283.00000        6       336   641.77000
18284.00000        1       494   461.68000
18285.00000        1       145   427.00000
18286.00000        2       608  1296.43000
18287.00000        4      1427  2345.71000

[4314 rows x 3 columns]

In [73]:
cltv_c.columns = ['total_transaction','total_unit','total_price']

In [75]:
cltv_c

total_transaction  total_unit  total_price
Customer ID                                            
12346.00000                 11          70    372.86000
12347.00000                  2         828   1323.32000
12348.00000                  1         373    222.16000
12349.00000                  3         993   2671.14000
12351.00000                  1         261    300.93000
...                        ...         ...          ...
18283.00000                  6         336    641.77000
18284.00000                  1         494    461.68000
18285.00000                  1         145    427.00000
18286.00000                  2         608   1296.43000
18287.00000                  4        1427   2345.71000

[4314 rows x 3 columns]

* Yukarıda customer ıd leri tekilleştirdik.Bir müşteri 11 tane işlem yapmış 70 birim işlem yapmış ve 372.86 bir harcama yapmış

<span style="color: blue; font-weight: bold;">2. AVERAGE ORDER VALUE</span> 

total_price / total_transaction

In [83]:
cltv_c['total_price'] / cltv_c['total_transaction']

Customer ID
12346.00000    33.89636
12347.00000   661.66000
12348.00000   222.16000
12349.00000   890.38000
12351.00000   300.93000
                 ...   
18283.00000   106.96167
18284.00000   461.68000
18285.00000   427.00000
18286.00000   648.21500
18287.00000   586.42750
Length: 4314, dtype: float64

In [85]:
cltv_c['average_order_value'] = cltv_c['total_price'] / cltv_c['total_transaction']

<span style="color: blue; font-weight: bold;">3. PURCHASE FREQUENCY</span>

satın alma sıklığı = total_transaction / total_number_of_customers

In [91]:
cltv_c.shape

(4314, 4)

total_number_of_customers:toplam müşteri sayısı demek, her ID bir müşteridir.
o yüzden shape[0] ile satır sayısılarına yani toplam müşteri sayısına erişirim.

In [96]:
cltv_c['Purchase_frequency'] = cltv_c['total_transaction'] / cltv_c.shape[0]

<span style="color: blue; font-weight: bold;">4.REPEAT RATE/ CHURN RATE</span>

Churn Rate = 1- Repaet Rate

Repeat Rate: Tekrar oranı, elde tutma oranıdır.

Repaet Rate = Birden fazla alışveriş yapan müşteri sayısı / tüm müşteriler

In [103]:
cltv_c[cltv_c['total_transaction']>1] #birden fazla alışveriş yapan müşteriler

total_transaction  total_unit  total_price  average_order_value  Purchase_frequency
Customer ID                                                                                     
12346.00000                 11          70    372.86000             33.89636             0.00255
12347.00000                  2         828   1323.32000            661.66000             0.00046
12349.00000                  3         993   2671.14000            890.38000             0.00070
12352.00000                  2         188    343.80000            171.90000             0.00046
12356.00000                  3        1826   3562.25000           1187.41667             0.00070
...                        ...         ...          ...                  ...                 ...
18276.00000                  5        1060   1320.66000            264.13200             0.00116
18277.00000                  4         381   1069.67000            267.41750             0.00093
18283.00000                  6         336    641.77000            106.96167             0.00139
18286.00000                  2         608   1296.43000            648.21500             0.00046
18287.00000                  4        1427   2345.71000            586.42750             0.00093

[2893 rows x 5 columns]

In [105]:
cltv_c[cltv_c['total_transaction']>1].shape[0]#birden fazla alışveriş yapan müşteri sayısı

2893

In [107]:
repeat_rate = cltv_c[cltv_c['total_transaction']>1].shape[0] / cltv_c.shape[0]

In [109]:
churn_rate = 1-repeat_rate

In [111]:
churn_rate

0.3293926750115902

<span style="color: blue; font-weight: bold;">5. PROFİT MARGİN</span>

Profit Margin = Total Price * 0.10 #burdaki sabit değişebilir.

In [118]:
cltv_c['profit_margin'] = cltv_c['total_price']*0.10

<span style="color: blue; font-weight: bold;">6. CUSTOMER VALUE</span>

Customer Value = Average Order Value * Purchase Frequency

In [125]:
cltv_c['customer_value'] = cltv_c['average_order_value'] *  cltv_c['Purchase_frequency']

In [129]:
cltv_c.head()

total_transaction  total_unit  total_price  average_order_value  Purchase_frequency  profit_margin  customer_value
Customer ID                                                                                                                    
12346.00000                 11          70    372.86000             33.89636             0.00255       37.28600         0.08643
12347.00000                  2         828   1323.32000            661.66000             0.00046      132.33200         0.30675
12348.00000                  1         373    222.16000            222.16000             0.00023       22.21600         0.05150
12349.00000                  3         993   2671.14000            890.38000             0.00070      267.11400         0.61918
12351.00000                  1         261    300.93000            300.93000             0.00023       30.09300         0.06976

<span style="color: blue; font-weight: bold;">7. CUSTOMER LIFETIME VALUE</span>

CLTV = (Customer Value / Churn Rate) * Profit Margin

In [133]:
cltv_c['cltv'] = (cltv_c['customer_value'] / churn_rate)* cltv_c['profit_margin']

In [135]:
cltv_c.head()

total_transaction  total_unit  total_price  average_order_value  Purchase_frequency  profit_margin  customer_value      cltv
Customer ID                                                                                                                              
12346.00000                 11          70    372.86000             33.89636             0.00255       37.28600         0.08643   9.78357
12347.00000                  2         828   1323.32000            661.66000             0.00046      132.33200         0.30675 123.23546
12348.00000                  1         373    222.16000            222.16000             0.00023       22.21600         0.05150   3.47326
12349.00000                  3         993   2671.14000            890.38000             0.00070      267.11400         0.61918 502.11041
12351.00000                  1         261    300.93000            300.93000             0.00023       30.09300         0.06976   6.37290

In [141]:
cltv_c.sort_values(by='cltv', ascending=False).head()

total_transaction  total_unit  total_price  average_order_value  Purchase_frequency  profit_margin  customer_value          cltv
Customer ID                                                                                                                                  
18102.00000                 89      124216 349164.35000           3923.19494             0.02063    34916.43500        80.93749 8579573.77276
14646.00000                 78      170342 248396.50000           3184.57051             0.01808    24839.65000        57.57916 4342070.45829
14156.00000                102      108107 196566.74000           1927.12490             0.02364    19656.67400        45.56484 2719105.08615
14911.00000                205       69722 152147.57000            742.18327             0.04752    15214.75700        35.26833 1629055.80978
13694.00000                 94      125893 131443.19000           1398.33181             0.02179    13144.31900        30.46898 1215855.89003

<span style="color: blue; font-weight: bold;">8.SEGMENTLERİN OLUŞTURULMASI</span>

In [145]:
cltv_c['segment'] = pd.qcut(cltv_c['cltv'],4, labels = ['D','C','B','A'])

In [147]:
cltv_c.sort_values(by='cltv', ascending=False).head()

total_transaction  total_unit  total_price  average_order_value  Purchase_frequency  profit_margin  customer_value          cltv segment
Customer ID                                                                                                                                          
18102.00000                 89      124216 349164.35000           3923.19494             0.02063    34916.43500        80.93749 8579573.77276       A
14646.00000                 78      170342 248396.50000           3184.57051             0.01808    24839.65000        57.57916 4342070.45829       A
14156.00000                102      108107 196566.74000           1927.12490             0.02364    19656.67400        45.56484 2719105.08615       A
14911.00000                205       69722 152147.57000            742.18327             0.04752    15214.75700        35.26833 1629055.80978       A
13694.00000                 94      125893 131443.19000           1398.33181             0.02179    13144.31900        30.46898 1215855.89003       A

In [149]:
cltv_c.groupby('segment').agg({'count','mean','sum'})

C:\Users\ASLI\AppData\Local\Temp\ipykernel_20780\2986875375.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  cltv_c.groupby('segment').agg({'count','mean','sum'})


total_transaction                total_unit                    total_price                  average_order_value                 Purchase_frequency               profit_margin                 customer_value                         cltv                  
                      sum     mean count        sum       mean count           sum       mean count                 sum      mean count                sum    mean count           sum      mean count            sum    mean count            sum        mean count
segment                                                                                                                                                                                                                                                             
D                    1326  1.22892  1079     117616  109.00463  1079  192265.13000  178.18826  1079        169805.46550 157.37300  1079            0.30737 0.00028  1079   19226.51300  17.81883  1079       44.56772 0.04130  1079     2849.32790     2.64071  1079
C                    2160  2.00371  1078     305135  283.05659  1078  513016.45300  475.89652  1078        317422.35514 294.45487  1078            0.50070 0.00046  1078   51301.64530  47.58965  1078      118.91897 0.11031  1078    18184.26342    16.86852  1078
B                    4063  3.76902  1078     733211  680.15863  1078 1219605.20000 1131.35918  1078        420127.12365 389.72832  1078            0.94182 0.00087  1078  121960.52000 113.13592  1078      282.70867 0.26225  1078   103549.86329    96.05739  1078
A                   11666 10.81186  1079    4383262 4062.33735  1079 6907116.49100 6401.40546  1079        723972.22347 670.96592  1079            2.70422 0.00251  1079  690711.64910 640.14055  1079     1601.09330 1.48387  1079 25257295.04118 23408.05843  1079

In [152]:
cltv_c.to_csv('cltv_c.csv')#oluşturulan segmentlerin dışarı aktarılması

<span style="color: blue; font-weight: bold;">9.TÜM İŞLEMLERİN FONKSİYONLAŞTIRILMASI</span>

In [ ]:
def create_cltv_c(dataframe, profit=0.10):

    # Veriyi hazırlama
    dataframe = dataframe[~dataframe["Invoice"].str.contains("C", na=False)]
    dataframe = dataframe[(dataframe['Quantity'] > 0)]
    dataframe.dropna(inplace=True)
    dataframe["TotalPrice"] = dataframe["Quantity"] * dataframe["Price"]
    cltv_c = dataframe.groupby('Customer ID').agg({'Invoice': lambda x: x.nunique(),
                                                   'Quantity': lambda x: x.sum(),
                                                   'TotalPrice': lambda x: x.sum()})
    cltv_c.columns = ['total_transaction', 'total_unit', 'total_price']
    # avg_order_value
    cltv_c['avg_order_value'] = cltv_c['total_price'] / cltv_c['total_transaction']
    # purchase_frequency
    cltv_c["purchase_frequency"] = cltv_c['total_transaction'] / cltv_c.shape[0]
    # repeat rate & churn rate
    repeat_rate = cltv_c[cltv_c.total_transaction > 1].shape[0] / cltv_c.shape[0]
    churn_rate = 1 - repeat_rate
    # profit_margin
    cltv_c['profit_margin'] = cltv_c['total_price'] * profit
    # Customer Value
    cltv_c['customer_value'] = (cltv_c['avg_order_value'] * cltv_c["purchase_frequency"])
    # Customer Lifetime Value
    cltv_c['cltv'] = (cltv_c['customer_value'] / churn_rate) * cltv_c['profit_margin']
    # Segment
    cltv_c["segment"] = pd.qcut(cltv_c["cltv"], 4, labels=["D", "C", "B", "A"])

    return cltv_c